doc/developer/design/20230717_alter_set_cluster.md

```diff
@@ -0,0 +1,108 @@
+# Moving source and sinks
+
+Issue: [#17417](https://github.com/MaterializeInc/materialize/issues/17417)
+
+## Context
+
+Materialize's API today allows users to specify a target cluster for sources
+and sinks on creation, but does not allow for changing the cluster at a later
+time. To change the cluster, users have to drop and recreate the object.
+
+This is problematic:
+* Sources almost always have downstream dependencies that need to be recreated,
+  too.
+* Sinks require users to recreate the topic in the downstream Kafka broker.
+
+## Goals
+
+* Users chan change the cluster on which a source or sink runs.
+
+## Non-Goals
+
+* Mixing compute and storage workload on the same cluster.
+
+## Overview
+
+We'll add support for the following declarative source and sink management
+commands to Materialize:
+
+```sql
+-- Move a source to cluster `clstr`:
+ALTER SOURCE src SET CLUSTER clstr;
+
+-- Move a sink to cluster `c2`:
+ALTER SINK snk SET CLUSTER c2;
+```
+
+Note that this is _intentionally_ lacking parens
+(`ALTER SOURCE src SET (CLUSTER = clstr)`;).
+This is for consistency with the analogous PostgreSQL command
+`ALTER TABLE ... SET {SCHEMA|TABLESPACE} {schema|tblspc}`.
+The rule is that only parameters configured in the `WITH` block go inside the
+parens; parameters that have dedicated syntax, like schemas, tablespaces, and
+clusters, get dedicated `ALTER ... SET` syntax too.
+
+We could offer an alternative syntax using `IN`:
+```sql
+ALTER SOURCE src IN CLUSTER clstr;
+```
+
+## Detailed description
+
+The `ALTER SOURCE` and `ALTER SINK` statements will learn a new option `SET CLUSTER clstr`:
+* The cluster name `clstr` must match an existing cluster.
+* The usual constraints apply. At the moment, the cluster must either be empty
+  or only host sources and sinks. The replication factor must be 0 or 1.
+* Clusters can be managed or unmanaged.
+
+When the command updates the cluster, Materialize will immediately drop the
```

Comment:
Could you add some more details around how dependencies will be handled? In the beginning of the doc, you mention that one limitation of dropping and recreating the objects is that dependencies need to be dropped too, but it isn't clear how this approach will work around that.

Reply:
I imagine this is "drop" from the cluster's perspective, but not the user's perspective. After all, the whole goal of this work is to let users move sources between clusters without having to drop and recreate their entire stack of downstream views and indexes.

Reply:
Yeah; I was looking to learn a bit more about what happens under the hood when you "drop a source/sink from a cluster's perspective".